### PR TITLE
bpo-28557: error message for bad raw readinto

### PR DESCRIPTION
**Pull request by davidszotten**
https://bugs.python.org/issue28557

### CODE REVIEW

**Comment by davidszotten** [comment]
Would prefer to still return an OSError with the TypeError as a cause, but couldn't figure out how to do that. Hints/help would be appreciated.

File: Modules/_io/bufferedio.c [hunk 1]
```diff
@@ -1488,6 +1488,14 @@ _bufferedreader_raw_read(buffered *self, char *start, Py_ssize_t len)
     }
     n = PyNumber_AsSsize_t(res, PyExc_ValueError);
     Py_DECREF(res);
+
+    if (n == -1 && PyErr_Occurred()) {
```
Comment (benjaminp) [inline Modules/_io/bufferedio.c:1487]: Why did you make a new if statement? Can't you put this logic in the next one?

File: Modules/_io/bufferedio.c [hunk 2]
```diff
@@ -1488,6 +1488,14 @@ _bufferedreader_raw_read(buffered *self, char *start, Py_ssize_t len)
...
+        _PyErr_TrySetFromCause(
```
Comment (benjaminp) [inline Modules/_io/bufferedio.c]: I think you actually want `_PyErr_FormatFromCause`?

File: Lib/test/test_io.py [hunk 3]
```diff
@@ -1531,6 +1531,22 @@ def test_args_error(self):
         with self.assertRaisesRegex(TypeError, "BufferedReader"):
             self.tp(io.BytesIO(), 1024, 1024, 1024)
 
+    def test_bad_readinto_value(self):
```
Comment (benjaminp) [inline Lib/test/test_io.py:1590]: You should put these tests in `BufferedReaderTest` and fix `_pyio.py`, too.

**Review by benjaminp — CHANGES_REQUESTED** [review]
Yeah, I don't think you should change the exception type.

File: Modules/_io/bufferedio.c (hunk 1 above)
Comment (davidszotten) [reply Modules/_io/bufferedio.c:1487]: not sure what you mean here. i may be wrong but i think i need a new if statement to check for this particular failure case.

or are you just suggesting i nest this if statement inside the `if (n < 0 || n > len)`?

Comment (benjaminp) [reply Modules/_io/bufferedio.c:1487]: I was thinking you could simply change the `PyErr_Format` on line 1493 to `_PyErr_FormatFromCause` and the the exception from `PyNumber_AsSsize_t` would be set as the cause (if there was one).

Comment (davidszotten) [reply Modules/_io/bufferedio.c:1487]: ah, i see. didn't realise that worked even if there was no exception. this is one of my first forays in to cpython so my knowledge level is a bit limited i'm afraid

Comment (davidszotten) [reply Modules/_io/bufferedio.c:1487]: hm, i may be doing something else wrong, but i think calling `_PyErr_FormatFromCause` with no current exception causes a segfault

File: Lib/test/test_io.py (hunk 3 above)
Comment (davidszotten) [reply Lib/test/test_io.py:1590]: i may be missing something but i can't see anything exactly analogous to this in the python version

Comment (benjaminp) [reply Lib/test/test_io.py:1590]: Okay, we can leave it then.

File: Modules/_io/bufferedio.c (hunk 1 above)
Comment (benjaminp) [reply Modules/_io/bufferedio.c:1487]: Alright, I was wrong to try to be clever.

**Comment by benjaminp** [comment]
I'm saying the error handling code should look like [this](https://github.com/python/cpython/blob/9eb35ab0d71a6bd680e84fa0f828cb634e72b681/Modules/_io/bufferedio.c#L711-L717).

**Comment by davidszotten** [comment]
friendly bump @benjaminp

**Comment by davidszotten** [comment]
@csabella should we change this from awaiting changes to awaiting review again?

File: Lib/test/test_io.py [hunk 4]
```diff
@@ -1574,6 +1574,22 @@ def test_args_error(self):
...
+        rawio = io.BufferedReader(io.BytesIO(b"12"))
+        rawio.readinto = lambda buf: -1
+        bufio = self.tp(rawio)
+        with self.assertRaises(OSError) as cm:
+            bufio.readline()
+        assert cm.exception.__cause__ is None
```
Comment (remilapeyre) [inline Lib/test/test_io.py]:
```suggestion
        self.assertIsNone(cm.exception.__cause__)
```

File: Lib/test/test_io.py [hunk 5]
```diff
@@ -1574,6 +1574,22 @@ def test_args_error(self):
...
+
+    def test_bad_readinto_type(self):
+        rawio = io.BufferedReader(io.BytesIO(b"12"))
+        rawio.readinto = lambda buf: b''
+        bufio = self.tp(rawio)
+        with self.assertRaises(OSError) as cm:
+            bufio.readline()
+        assert isinstance(cm.exception.__cause__, TypeError)
```
Comment (remilapeyre) [inline Lib/test/test_io.py]:
```suggestion
        self.assertIsInstance(cm.exception.__cause__, TypeError)
```

**Review by remilapeyre — CHANGES_REQUESTED** [review]

**Comment by davidszotten** [comment]
thanks for the suggestions @remilapeyre . updated

File: Modules/_io/bufferedio.c [hunk 6]
```diff
@@ -1483,6 +1483,15 @@ _bufferedreader_raw_read(buffered *self, char *start, Py_ssize_t len)
     }
     n = PyNumber_AsSsize_t(res, PyExc_ValueError);
     Py_DECREF(res);
+
+    if (n == -1 && PyErr_Occurred()) {
```
Comment (remilapeyre) [inline Modules/_io/bufferedio.c:1487]:
```suggestion
    if (n == -1) {
```

**Review by remilapeyre — APPROVED** [review]
Thanks @davidszotten ! Just one small nit, I'm pretty sure an error must be set when `n == 1` so you don't have to check `PyErr_Occurred()` too. Otherwise it looks good to me!

File: Modules/_io/bufferedio.c (hunk 6 above)
Comment (davidszotten) [reply Modules/_io/bufferedio.c:1487]: i don't quite understand why, but removing that check i get a segfault

here's the bottom of the stack from lldb:
```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x00000001001542f5 python.exe`_PyErr_FormatVFromCause [inlined] _Py_DECREF(op=0x0000000000000000) at object.h:441:9 [opt]
    frame #1: 0x00000001001542f5 python.exe`_PyErr_FormatVFromCause(tstate=0x0000000100506680, exception=0x0000000100292fc0, format="raw readinto() returned invalid value", vargs=0x00007ffeefbf9a40) at errors.c:586 [opt]
  * frame #2: 0x0000000100154460 python.exe`_PyErr_FormatFromCause(exception=<unavailable>, format=<unavailable>) at errors.c:626:5 [opt]
    frame #3: 0x00000001001e8508 python.exe`_bufferedreader_raw_read(self=<unavailable>, start=<unavailable>, len=8192) at bufferedio.c:1488:9 [opt]
```

Comment (davidszotten) [reply Modules/_io/bufferedio.c:1487]: hang on, i think i get it now.  the point is that `n` is the return value of the (user-defined) `readinto` which means it _can_ sometimes be `-1` even without `PyErr_Occurred`. this is then caught in the next `if` block (and becomes "raw readinto() returned invalid length [..]")

the whole point of this pr is to help users debug bad `readinto` implementations by providing more granular error messages

Comment (remilapeyre) [reply Modules/_io/bufferedio.c:1487]: Yes sorry, this was a mistake :/

**Comment by davidszotten** [comment]
@remilapeyre no worries

does that mean this is ready? pr is marked as "requested changes" from @benjaminp though he never replied to my [comment about his suggested change](https://github.com/python/cpython/pull/7496#issuecomment-531351841)

what's the next step?

**Comment by remilapeyre** [comment]
> @remilapeyre no worries
> 
> does that mean this is ready? pr is marked as "requested changes" from @benjaminp though he never replied to my [comment about his suggested change](https://github.com/python/cpython/pull/7496#issuecomment-531351841)
> 
> what's the next step?

I think your change looks good, new it needs to be reviewed by a core developer like @benjaminp  and they will either merge it or ask you for changes. There is a lot of work going on on Python and they have  a lot of PR to review so it may take them some time but they will get there :)

**Comment by davidszotten** [comment]
is there a difference between the tags for "awaiting review" and "awaiting core review" that's applicable here?

**Review by remilapeyre — APPROVED** [review]
The first is when nobody reviewed, the second is when someone who is not a core developer has made a review. It should have switched but the bot seems to have missed it.

**Comment by davidszotten** [comment]
thank you!

**Pull request merged**